sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -110,23 +110,28 @@ object ExpressionEncoder {
     }
     val newSerializer = CreateStruct(serializers)
 
+    def nullSafe(input: Expression, result: Expression): Expression = {
+      If(IsNull(input), Literal.create(null, result.dataType), result)
+    }
+
     val newDeserializerInput = GetColumnByOrdinal(0, newSerializer.dataType)
     val deserializers = encoders.zipWithIndex.map { case (enc, index) =>
       val getColExprs = enc.objDeserializer.collect { case c: GetColumnByOrdinal => c }.distinct
       assert(getColExprs.size == 1, "object deserializer should have only one " +
         s"`GetColumnByOrdinal`, but there are ${getColExprs.size}")
 
       val input = GetStructField(newDeserializerInput, index)
-      enc.objDeserializer.transformUp {
+      val newDeserializer = enc.objDeserializer.transformUp {
```

Comment:
do you mean the `objDeserializer` can be something that doesn't propagate nulls, so we need to manually check null input and create null literal? If so please add some code comments to explain it.

Reply:
It seems so. I'm new to this part of the code so you'll certainly have a better idea of what's going on here.
Could it happen because `joinWith` creates a tuple from what were top-level Rows?

Reply:
Can you run the test locally, add some print here to see what's the `objDeserializer` that causes the bug?

Reply:
Oh sure! It's a CreateExternalRow Expression. If I `println(enc.objDeserializer)`, it prints
```
createexternalrow(getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._0.toString, getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._1, StructField(a,StringType,true), StructField(b,IntegerType,false))
createexternalrow(getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._0.toString, getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._1, StructField(a,StringType,true), StructField(b,IntegerType,false))
```

Reply:
Unless there's a broader change to make, we could reduce the blast radius by:
- limiting the change to `CreateExternalRow` (i.e. check `enc.objDeserializer.isInstanceOf[CreateExternalRow]`)?
- having a dedicated tuple ExpressionEncoder for `Dataset.joinWith` (i.e. ~ update `ExpressionEncoder.tuple` to add a `nullSafe: Boolean = false` flag, set it to true for `Dataset.joinWith` and manually propagate nulls if true) ?

Reply:
> having a dedicated tuple ExpressionEncoder for Dataset.joinWith

I like this idea. Ideally, for an `ExpressionEncoder[T]`, the deserialize expression assumes the input row is not null, as the input rows are from a `DataFrame` and Spark doesn't allow null as top-level rows. We can see from `ScalaReflection.deserializerForType` that we use `AssertNotNull` to guarantee it. This null check seems missing in `RowEncoder.deserializerFor`

In `ExpressionEncoder.tuple`, we inherit this assumption and assume each element of the Tuple can't be null. This assumption is broken in `joinWith`, because the Tuple element can be null for outer joins.

Reply:
This bug only occurs for `RowEncoder`, as `Dataset[T].joinWith(Dataset[U])` works fine: https://github.com/apache/spark/pull/13425/files#diff-b98c99535d2b28cb47774860d500030e732c244c55b1ac05aead5d1cf1e7a602R772

Can you look into it and figure out the difference? This may help us the understand the bug better and guide us to the proper fix.

Reply:
The difference comes from the `RowEncoder` deserializer:
- When using a `Dataset[T]`, the `ExpressionEncoder` is used and calls [`ScalaReflection.deserializerForType`](https://github.com/apache/spark/blob/edc52857ca3ac031481a910f7871ff8d5e030e18/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L56) to get a deserializer for class `T`, which automatically [wraps the expression in a null-safe expression](https://github.com/apache/spark/blob/58d3f1516ed812b692709991e551829aa0090578/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L390-L394).
- When using a `DataFrame`, the `RowEncoder` is used and returns a [`CreateExternalRow` (not wrapped in a null-safe expression)](https://github.com/apache/spark/blob/58d3f1516ed812b692709991e551829aa0090578/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L254-L259).

I'm not sure there's an easy way to solve this, as the RowEncoder should guarantee (afaik) that top-level Rows aren't `null`.

Actually I think everything is already summarized in your initial PR that patched the tuple encoder to wrap deserializers in a null-safe way: https://github.com/apache/spark/pull/13425

Reply:
> as the RowEncoder should guarantee (afaik) that top-level Rows aren't null.

This is not true for outer join. Shall we also add a null check to wrap `CreateExternalRow`?

Reply:
OK let's update `joinWith` then.